rpl: introducing P2P-RPL

cgundogan opened the pull request:
This PR introduces [P2P-RPL](https://tools.ietf.org/html/rfc6997) as a separate module to `gnrc_rpl`.

Can be tested by including `gnrc_rpl_p2p` in the `gnrc_networking` example Makefile.
After manually setting up some nodes with an ipv6 address (same prefix) (or by using another DODAG for SLAAC), then a route can be discovered by issuing:
 `rpl find <ip_src> <ip_dst>`
`<ip_src>` resembles a configured ipv6 address to be used as the `DODAG-ID`
`<ip_dst>` is the ipv6 address that is being searched in the network

This PR only includes the feature to install Hop-by-Hop routes.
The possibility of finding source routes will be addressed when #3647 is merged.

Depends on ~~#3718~~ + ~~#3719~~ + ~~#3720~~

BytesGalore commented:
updated the label, please rebase

cgundogan commented:
I took this out of the 2015.12 release, because I won't be able to rebase and test it again 'til next week

zhuoshuguo commented:
@cgundogan Ping. A kind reminder for rebase.

cgundogan commented:
This is no longer WIP. I usually tested it like the following on native:
1. `gnrc_networking` Makefile: `gnrc_rpl_p2p` instead of `gnrc_rpl`. For testing `gnrc_ipv6_whitelist`
2. created 4 tap interfaces
3. assigned manually global ipv6 addresses to the nodes (`abcd::x`)
4. configured the proper whitelisting so that all tap interfaces form a line
5. called `rpl find abcd::1 abcd::4` to start a temporary DODAG with the DODAG-ID `abcd::1`. The target of the route discovery is `abcd::4`.
6. with the shell command `rpl` the temporary and short-lived DODAG can be verified.
7. after a small amount of time a FIB entry should appear for all nodes but the target, pointing from the DODAG root to the target

cgundogan commented:
A friendly ping to @BytesGalore (:

BytesGalore commented:
since #4732 is in, please rebase

BytesGalore commented:
maybe I overlooked something resolving the conflicts, but I get:
```
gnrc_rpl_control_messages.c:179:14: error: ‘gnrc_rpl_dodag_t’ has no member named ‘prefix_info_requested’
```
when I try to build `example/gnrc_networking` with `gnrc_rpl_p2p` instead of `gnrc_rpl`

BytesGalore commented:
Ahh I see now its changed in #4732 from `bool` to the flag `GNRC_RPL_REQ_OPT_PREFIX_INFO`

cgundogan commented:
hm you are right :/ will rebase in a bit

cgundogan commented:
@BytesGalore rebased

OlegHahm commented:
So?

BytesGalore commented:
had no chance for a descent look, will try on the weekend.

emmanuelsearch commented:
For some reason, I still get the error:
```
/home/vagrant/RIOT/sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:98:10: error: 'gnrc_rpl_dodag_t {aka struct gnrc_rpl_dodag}' has no member named 'dodag_conf_requested'
     dodag->dodag_conf_requested = true;
```

cgundogan commented:
@emmanuelsearch thank you for the hint! The issue should be fixed now.

emmanuelsearch commented:
Tested as proposed (4 nodes in a line). The temporary DAGs are formed, and the FIB entries do appear.
However, forwarding does not work: ping6 and udp send work between neighbors (using GUA or local address) but not whenever multi-hop is involved, though the FIB entries are there and seem correct. Any idea why? Did you try using the routes? Does it work on your setup?

cgundogan commented:
I tested on the iotlab testbed today on a multi-hop path and I was able to send udp packets from the originator (A) of the p2p DAG to the target (B) (after the finding process). The ping6 command won't work, because the path is built uni-directional from the originator (A) to the target (B). Ping6 should work if the same process is done in reversed order (A -> B and B -> A)

emmanuelsearch commented:
Weird. I tested exactly the same on native and it did not work (the way I described in my last comment). Can you check on native if it works for you?

cgundogan commented:
@emmanuelsearch just to be sure: did you also whitelist the p2p originator GUA at the p2p target before sending udp traffic?

Setting up all those whitelist entries is very cumbersome and prone to errors .. I remembered that I have this PR (https://github.com/des-testbed/desvirt/pull/17) lying around, so that RIOT can be used in desvirt again. This way we can set up a line topology without hassle. I just tested with that PR and I was able to send udp traffic from the p2p originator to the p2p target. If you are around later maybe we can tackle that together?

emmanuelsearch commented:
works on native after whitelisting things correctly ;)

emmanuelsearch commented:
I'll try to look into the code quickly in the next few days, to try and detect any logical issues or mismatch with the RFC.
Else we could aim to merge this new functionality soon, and refine/extend later if needed?

emmanuelsearch commented:
@authmillenon by the way, whitelist would be less confusing and less cumbersome for users if it offered a mode to filter at MAC layer too (instead of only at IP layer). What do you think?

authmillenon commented:
Without #4646 this is something every driver would need for itself (due to different MAC address formats). With #4646 this can be relegated to the glue code, I think.

BytesGalore commented on sys/include/net/gnrc/rpl/p2p.h:
```diff
@@ -0,0 +1,180 @@
+ *              RFC 6997, section 7, P2P Route Discovery Option (P2P-RDO)
+ *          </a>
+ */
+#define GNRC_RPL_P2P_OPT_RDO    (10)
```
I would vote for a hex representation `(0x0a)` as used in the RFC

BytesGalore commented on sys/include/net/gnrc/rpl/p2p.h:
```diff
@@ -0,0 +1,180 @@
+gnrc_pktsnip_t *gnrc_rpl_p2p_rdo_build(gnrc_pktsnip_t *pkt, gnrc_rpl_p2p_ext_t *p2p_ext);
+
+/**
+ * @brief Parse an RDO
```
didn't looked into the implementation yet, but I guess the RDO will be also processed in the function not just parsed

BytesGalore commented on sys/include/net/gnrc/rpl/p2p_structs.h:
```diff
@@ -0,0 +1,106 @@
+    bool reply;             /**< request P2P-DRO */
+    bool dro_ack;           /**< request P2P-DRO-ACK */
+    uint8_t lifetime_enc;   /**< encoded lifetime of the P2P-DODAG */
+    int8_t lifetime_sec;    /**< lifetime of the P2P-DODAG in seconds */
```
signed?

BytesGalore replied:
ok, no objection since the DODAG lifetime is limited to 1 min max.

BytesGalore commented on sys/include/net/gnrc/rpl/p2p_structs.h:
```diff
@@ -0,0 +1,106 @@
+    bool for_me;            /**< true if this node is the target */
+    uint8_t addr_numof;     /**< number of addresses in the address vector */
+    int8_t dro_delay;       /**< delay DRO after it was requested in seconds */
+    ipv6_addr_t addr_vec[GNRC_RPL_P2P_ADDR_VEC_NUMOF];   /**< address vector */
```
I guess I will run into it further donw, but what addresses are stored here?

BytesGalore replied:
so `GNRC_RPL_P2P_ADDR_VEC_NUMOF`+1 is the max hop-count for P2P communication.
Please add this to the comment.

cgundogan replied:
addressed

BytesGalore commented on sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c:
```diff
@@ -153,6 +177,7 @@ void gnrc_rpl_dodag_remove_all_parents(gnrc_rpl_dodag_t *dodag)
     LL_FOREACH_SAFE(dodag->parents, elt, tmp) {
         gnrc_rpl_parent_remove(elt);
     }
+    dodag->my_rank = GNRC_RPL_INFINITE_RANK;
```
It doesn't hurt, but when we have no parents we will leave the DODAG anyway.

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+
+#define GNRC_RPL_P2P_RDO_LEN            (18)
+#define GNRC_RPL_P2P_RDO_FLAGS_LIFETIME (6)
+#define GNRC_RPL_P2P_RDO_FLAGS_HbH      (6)
```
I'm against lower case in defines, even for abbreviating the word `by`

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+    rdo->length = GNRC_RPL_P2P_RDO_LEN + addr_size;
+    rdo->compr_flags = (p2p_ext->hop_by_hop << GNRC_RPL_P2P_RDO_FLAGS_HbH) |
+                       (p2p_ext->compr & GNRC_RPL_P2P_RDO_FLAGS_COMPR);
+    rdo->lmn = (((rdo->length + 2 - sizeof(*rdo)) / addr_len) & GNRC_RPL_P2P_RDO_FLAGS_NEXT_HOP);
```
The infamous magic `2` ;)
Would be cool if you could state a comment here.

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+    }
+
+    flags = byteorder_ntohs(dro->flags_rev);
+    p2p_ext->stop = (flags & (1 << GNRC_RPL_P2P_DRO_FLAGS_STOP)) >> GNRC_RPL_P2P_DRO_FLAGS_STOP;
```
I wonder if keeping the flags as one `flags` member of the struct  would be more compact/readable.

cgundogan replied:
Could be more efficient, but I think less readable. Let's save this work for a follow-up PR.

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+    rdo = rdo_snip->data;
+
+    if (rdo->lmn > 0) {
+        memcpy(&addr.u8[p2p_ext->compr], &addr_vec[addr_len * (--rdo->lmn)], addr_len);
```
ughh `&addr_vec[addr_len * (--rdo->lmn)]`

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p_dodag.c:
```diff
@@ -0,0 +1,63 @@
+            gnrc_rpl_p2p_exts[i].state = 1;
+            gnrc_rpl_p2p_exts[i].dodag = dodag;
+            gnrc_rpl_p2p_exts[i].dro_delay = -1;
+            gnrc_rpl_p2p_exts[i].lifetime_sec = INT8_MIN;
```
shouldn't we use `GNRC_RPL_P2P_DEFAULT_LIFETIME` for endless/invalid?

cgundogan replied:
`GNRC_RPL_P2P_DEFAULT_LIFETIME` is used by the root node. All normal nodes will read the proper value (which is in our case also `GNRC_RPL_P2P_DEFAULT_LIFETIME`) from the RDO.

cgundogan commented:
>ok, first static code sight done. More to come :)
So far so good :+1:

thanks so much @BytesGalore  (: I will address them this evening or tomorrow morning

cgundogan commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+    dro = opt_snip->data;
+    dro->instance_id = p2p_ext->dodag->instance->id;
+    dro->version_number = 0;
+    dro->flags_rev = byteorder_htons((((p2p_ext->stop << 1) | (p2p_ext->dro_ack << 0))
```
yeah that's true, but I like to stress that the flag is at bit position 0 if someone looks at the packet format in the RFC (:

BytesGalore replied:
ok

cgundogan commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+    uint8_t *addr = NULL;
+    uint8_t addr_len = sizeof(ipv6_addr_t) - p2p_ext->compr;
+    uint8_t i = 0;
+    for (i = 0; i < addr_num; i++) {
```
hm, I tried that approach but it was not so trivial (: Reverted the change for now.

cgundogan commented:
@BytesGalore addressed your comments in 3a9217a

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG && defined(MODULE_IPV6_ADDR)
```
I just stuck on this one, `&& defined(MODULE_IPV6_ADDR)` is always `true` for this implementation, since you use `ipv6_addr_t`

cgundogan commented:
BTW, also rebased on master.
Can now be tested as follows with the `gnrc_networking` example:
1. change `gnrc_rpl` to `gnrc_rpl_p2p` in the `Makefile`
2. TOPO_TYPE=line TOPO_SIZE=4 make desvirt-topology
3. TOPO=line4 make desvirt-define
4. TOPO=line4 make clean all desvirt-start
5. in 4 different consoles:
  1. `nc localhost 4712` (leftmost node in the line)
  2. `nc localhost 4711`
  3. `nc localhost 4714`
  4. `nc localhost 4713` (rightmost node in the line)

6.`rpl init 6` on all nodes and `ifconfig 6 add abcd::x` (1-4)
7.`rpl find abcd::1 abcd::4` on the leftmost node
8. `rpl` to see the short-lived DODAG and `fibroute` on all nodes to verify the entries
9. `udp server start 8888` on the rightmost node
10. `udp send abcd::4 8888 hello` on the leftmost node
11. the rightmost node should receive and print the message
12. TOPO=line4 make desvirt-stop

BytesGalore commented on sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c:
```diff
@@ -313,6 +348,10 @@ static gnrc_rpl_parent_t *_gnrc_rpl_find_preferred_parent(gnrc_rpl_dodag_t *doda
+    }
+#endif
+
     dodag->my_rank = dodag->instance->of->calc_rank(dodag->parents, 0);
```
I wonder if our rank can change when we are in a P2P DODAG?
I'm asking since the preferred parent does not change and in this case we can just return the current preferred parent at L316.

cgundogan replied:
hmm this seems wrong as it currently is. The preferred parent should be able to change

cgundogan replied:
@BytesGalore I addressed this issue. Could you have a look at commit https://github.com/RIOT-OS/RIOT/pull/3721/commits/14fe84de7d72da0e0c051fa475fbf240b339a8d7 ?

BytesGalore replied:
@cgundogan currently I cannot test it, but from the code side it looks as we are now able to compute and choose a preferred-parent int the p2p DODAG :+1:

BytesGalore commented:
I guess I was just distracted while going thru the code again, could you point me to where you ensure the addresses in RDO/DRO messages for the path are loop-free?

cgundogan commented:
>I guess I was just distracted while going thru the code again, could you point me to where you ensure the addresses in RDO/DRO messages for the path are loop-free?

sure, look at `/dev/null` (: jk aside, I did not implement this for now, because in our case we only have one interface per node. But I will add this as a follow up PR, probably the same way as I did it for the source routing header.

BytesGalore commented:
ok, but we have to watch out then to also add a check for duplicates when an intermediate hop adds itself to the address vector of a DRO [1].

[1] https://github.com/RIOT-OS/RIOT/pull/3721/files#diff-5a48370fe78f64fe6328671560c0e53aR203

cgundogan commented:
yes, maybe we can factor this out somehow from [1]. Not sure though if they are both compatible.

[1] https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/routing/rpl/srh/gnrc_rpl_srh.c#L65

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+        rdo->lmn--;
+        memcpy(&addr.u8[p2p_ext->compr], &addr_vec[addr_len * rdo->lmn], addr_len);
+    }
+
```
We should also consider if we return here early if `GNRC_RPL_P2P_RDO_FLAGS_HBH` is set.
( Somehow like a reminder when Source Routes are come in (: )

cgundogan replied:
I don't quite understand why we should return early? The HBH route is set in the fib below that line.

BytesGalore replied:
That was exactly the point where my mind-twists started :)
I read section 9.1 in rfc6997 and thought we shouldn't add a FIB entry for a hop-by-hop state to avoid interference.

BytesGalore commented on sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c:
```diff
@@ -0,0 +1,390 @@
+    if_id = gnrc_ipv6_netif_find_by_addr(&me, &addr);
+
+    if (if_id != KERNEL_PID_UNDEF) {
+        fib_add_entry(&gnrc_ipv6_fib_table, if_id, p2p_ext->target.u8,
```
This one gives me serious mind-twists.
I think we must avoid to interfere with the genuine RPL controlled routing-domain here [1].
Or we end up with an inconsistent topology in the non-p2p DODAG.
On the other hand we must somehow set a route for p2p hop-by-hop and use it in favour of the genuine (rfc6550) DODAG. 
I mean if I ask the FIB for a next-hop to forward data and we have a p2p route, I would expect to receive this (rfc6997) next-hop and not one towards the genuine (rfc6550) DODAG root node.
But I have no clue how we could achieve this, since RPL is not asked for a next-hop but the FIB.
So for my understanding we are forced to somehow suppress/overrule the FIB for the p2p hop-by-hop lifetime. 
But this sounds just wrong or like a hack.

[1] https://tools.ietf.org/html/rfc6997#section-9.1

cgundogan replied:
what if we set it not as an `FIB_FLAG_RPL_ROUTE` but as an external route? This way those routes will be marked in the DAO as external.

BytesGalore replied:
I think this wouldn't prevent to overwrite stored entries in the FIB from a previously set downward route. For my understanding, the prefix and addresses of the nodes won't change in the temporary 6997 DODAG, they are still the same nodes within the 6550 DODAG.
In the end we would have inconsistencies in the 6550 topology.
Or do I miss something obvious (:eyeglasses:)

cgundogan replied:
Do we have a concept of priorites in the fib, or is that simple to add? We could do e.g. something like `fib_add_entry(...,2)` (with prio=2) and if there is an entry already inside the fib, the route will only be added if the current prio is > 2. This way normal "RPL" could have a higher prio and may override p2p-routes, but not the other way around.

BytesGalore replied:
no, there's no prio concept for the FIB, since the one that adds entries to the FIB doesn't know about others doing the same. For this case it would work of course, since p2p knows of 6550 RPL.
But I'm against adding ambiguous entries to the FIB, even if they can be dispatched by the provided additional parameter.
I have no specific idea how, but we could "take-out" the colliding FIB entry until the p2p lifetime expires, and re-add the "taken-out" entry again it has lifetime left.
But obviously we have the same problem the other way around, when now a new DAO downward route overwrites the p2p hop (aarrrrgh).

emmanuelsearch commented:
Actually, the original idea behind RFC6997 was to provide protocols to:

1.  measure if the current RPL path was satisfactory (with a mechanism described in [RFC6998](https://tools.ietf.org/html/rfc6998)) and, if not,

2. discover alternative paths which might satisfy us better (with RFC6997)

3. be able to function standalone, even if basic RPL is not running

So in that regard, satisfactory P2P routes are more likely to be of _higher_ priority than paths installed by basic RPL.

emmanuelsearch commented:
Concerning multiple routing protocols manipulating a single FIB: this is a known issue, independent of RPL vs P2P-RPL. For example, it's solved on traditional routers (e.g. Cisco boxes) by a configured [administrative distance](https://en.wikipedia.org/wiki/Administrative_distance) associated to each protocol (and thus, associated to each entry manipulated by the protocol). For multiple entries with same length of prefix match, the smallest administrative distance is chosen as next hop. Something like this could be used?

waehlisch commented:
Usually, you distinguish between FIB and RIB, and the FIB includes a single entry per prefix, based on a selection process, which might consider administrative distance.

Anyhow, I'm more confused about  ""static route" < "P2P-RPL" < "RPL" < "Default"". A Default route specifies an address destination. It is set per static configuration or (maybe) distributed via routing protocol.

emmanuelsearch commented:
Here, we are talking about multiple RIBs and a single FIB. 

If we want a single FIB entry per "destination", we need additional logic+state somewhere, through which routing protocols would indirectly manage FIB manipulation, right? So the question is, how to optimize the amount of this additional logic+state. 

Pre-configured administrative distance per protocol running, and administrative distance field in the FIB is a simple way to do that. What are the potentially more efficient alternatives, in practice?

BytesGalore commented:
I'm with @waehlisch, the FIB should not do any administrative extra operations, a "parent process" such as a RIB or RPL should do this.

BytesGalore commented:
>Pre-configured administrative distance per protocol running, and administrative distance field in the FIB is a simple way to do that. What are the potentially more efficient alternatives, in practice?

But how would RPL know of other routing protocols or different instances of routing protocols per interface?
And how would we ensure that a discovered route is more preferable over the other?

OlegHahm commented:
> a "parent process" such as a RIB or RPL should do this.

Wouldn't that required that potentially co-existing routing protocols would need to cooperate?

BytesGalore commented:
I mean "someone" needs to add the administrative distance and needs knowledge about all protocols running.

emmanuelsearch commented:
@BytesGalore yes, but this assumption is realistic. Someone _has_ to know what routing protocols were flashed on the device anyways, and how they _should_ interplay.

BytesGalore commented:
Per default we would usually never run two protocols for the same prefix, because of routing loops.
For this special case here, where basically two protocols (rfc6550 vs rfc6997) operate an equal prefix, we could flag the rfc6997 FIB entries and not overwrite them if a new rfc6550 DAO route is about to be formed, so p2p always overrules rfc6550 DAO target routes.
RPL would read the next-hop and consider on the flags if it changes the next-hop for the target.

emmanuelsearch commented:
Wherever you put the logic+state (FIB or other process) to set priorities on paths, the (hard) problem remains the same for _ensuring_ this setting produces the desired results. 

Cooperation between routing protocols is a nightmare, and unrealistic or undoable in general
(hence the introduction of simplifying generic priority assumptions).

waehlisch commented:
@OlegHahm, no, co-existing routing protocols don't need to cooperate. There is a separate logic (aka route selection, or best path selection).

Yes, @emmanuelsearch, "administrative distance field in the FIB is a simple way" of implementation. However, it's always a bit confusing when we talk about FIB, and the (RIOT) FIB implements a mixture of RIB and FIB -- that's not the common understanding of a FIB.

More importantly, FIB is usually optimized for fast lookups of the destination addresses, where RIBs are optimized for fast updates (in RIB you have to analyse all covering prefixes anyway).

BytesGalore commented:
> Cooperation between routing protocols is a nightmare, and unrealistic or undoable in general
(hence the introduction of simplifying generic priority assumptions). 

I feel it :) right now.
As @waehlisch suggests, I would like to keep the operation of the FIB and the routing-decision-prefiltering-and-weighting-operations in separated processes specific for their purpose.
In this case the RPL process knows of rfc6550 and rfc6997 running equally on the node and has all necessary information to act as "someone" to decide if one learned next-hop should overrule the other.

emmanuelsearch commented:
@BytesGalore the solution you propose seems to make sense for RPL. But the first question that needs to be decided is: do we or do we not consider the case where multiple routing protocols (not necessarily RPL) run on the board, do not cooperate, and must come up with a single FIB? If yes, then an approach not tied to RPL would be more useful in the end (but we could decide to postpone for this).

cgundogan commented:
>In this case the RPL process knows of rfc6550 and rfc6997 running equally on the node and has all necessary information to act as "someone" to decide if one learned next-hop should overrule the other.

That's not quite true. We may know that we have multiple instances running, but as soon as a route finds its way into the FIB, I lose all RPL-related information to that route. So it's not possible to infer from the fib entry whether it was added by `RPL instance 1` or `RPL instance 2`. Thus, we cannot have any kind of priorisation on this level: "Am I able to override this route in the fib, or not.."

BytesGalore commented:
@emmanuelsearch I see the need of a RIB, I have a feeling that a RIB can be very complex learning and communicating with all protocols and apply policies on proposed routing decisions, so I just prefer it is separated from the FIB :)

@cgundogan if we flag the p2p FIB entries different to the usual 6550 entries, we can ask the FIB for the next-hop and see on the flag which RPL-"part" added it to the FIB. Then we can consider adding the new learned next-hop, or not.
If we have multiple RPL instances for the same prefix (which I would violently try to prevent, cause of nightmares) then you're of course right and the flag wouldn't help anything.

cgundogan commented:
Adding prioritisation information to FIB entries IMO does not break the concept of a "lightweight" FIB. Forwarding lookups will do the normal prefix matching without looking at priorities, so it won't change that part of the FIB. but the FIB itself is a central place for all routing protocols on the system.

I actuelly would go for something like:
```
fib_add_entry(...,...,... prio);
```
Any routing or "sub-routing (RPL-instances)" procotols can be assigned with a static priority that they put in as an argument. The fib can then ignore or accept this entry by just looking at the current prio of the same entry (if available).

This goes into the direction of the administrative distance that @emmanuelsearch mentioned.
However, it probably won't be possible (or it's very difficult) to change the priorities dynamically on run-time, because we again have no link from the fib entries to the routing protocols.

BytesGalore commented:
@cgundogan @emmanuelsearch I would vote for a lightweight solution for this case, e.g. the flags I mentioned.
Then we should thoroughly define and design a RIB concept for RIOT, before we start implement it :).

waehlisch commented:
Just one (obvious) clarification: If you use a single (RIB or in your words FIB) data structure to store routes in a multi routing protocol environment, you will most likely end in problems. Route decisions also depend on properties of the specific routing protocol. Administrative distance will be applied after this process has been passed. In the FIB you envision here, you lose the context.

emmanuelsearch commented:
OK, so what do we conclude here? We need to move on to finally merge this new RPL mode into master.

emmanuelsearch commented:
@cgundogan ping

waehlisch commented:
Can we reset our thoughts: In which scenario do we have multiple entries for the same prefix? If those scenarios do not appear in real deployment, we don't need to introduce complexity.

According to our discussions several months ago during network stack meeting with @kaspar030 and @tcschmidt, we excluded those scenarios.

emmanuelsearch commented:
As P2P-RPL can and does run standalone, we could indeed proceed as following:
- provide initial support for P2P-RPL running standalone only (single routing protocol case),
- defer to later the consideration of multiple routing protocols running concurrently
@cgundogan should we open an issue for multiple routing protocols manipulating the FIB?

cgundogan commented:
+1 for separating the multiple routing protocol issue from this PR.

@BytesGalore I think I addressed all your issues, could you have a quick look at them again?
@emmanuelsearch good plan, I will create that issue

tcschmidt commented:
concerning:

  "should we open an issue for multiple routing protocols manipulating 
the FIB?"

This should be no issue as long as one prefix/host is under the control 
of a single routing protocol, as discussed at length in spring (?) '15.

@waehlish asked for use cases otherwise - I don't see any in the IoT ??


On 18.03.2016 13:50, Emmanuel Baccelli wrote:
> As P2P-RPL can and does run standalone, we could indeed proceed as
> following:
>
>   * provide initial support for P2P-RPL running standalone only (single
>     routing protocol case),
>   * defer to later the consideration of multiple routing protocols
>     running concurrently @cgundogan <https://github.com/cgundogan>
>     should we open an issue for multiple routing protocols manipulating
>     the FIB?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3721#issuecomment-198338346>
>

waehlisch commented:
To conclude, @cgundogan please open  a new issue "Multiple routing instances for a single prefix". Here we can discuss scenarios, and if there are any, how to proceed.

BytesGalore commented:
@cgundogan so should we proceed using flags to dispatch 6997 and 6550 DAOs?

cgundogan commented:
rebased onto current master and squashed. @BytesGalore could you give it a quick test again?

>@cgundogan so should we proceed using flags to dispatch 6997 and 6550 DAOs?

let's move this to a follow-up PR, because we still need more discussion on this topic

BytesGalore commented:
I cannot test it right now. 
Can do this afternoon when I'm back from office.

> let's move this to a follow-up PR, because we still need more discussion on this topic

Indeed, that topic could include more workload then we imagine for now.

BytesGalore commented:
I fail to see why murdock failed

cgundogan commented:
apparently I have a doxygen error in p2p.h, but I have no clue what's the offending line

gebart commented:
@cgundogan run `make 2>&1 | less` in doc/doxygen and search: `/p2p.h`

cgundogan commented:
@BytesGalore I did some last-minute changes 9ed38092e3e50620381f153f6b75dc7e7911b70f

BytesGalore commented:
@cgundogan you can directly squash https://github.com/RIOT-OS/RIOT/commit/9ed38092e3e50620381f153f6b75dc7e7911b70f

cgundogan commented:
squashed

cgundogan commented:
@BytesGalore did you already ACK or did you want to test it again? Murdock seems happy now.

BytesGalore commented:
@cgundogan I cannot test it right now, so I can only ACK for the code not if it works properly.
I will test it this afternoon :)

BytesGalore commented:
@cgundogan can't get it running, neither with your desvirt description nor by manually seting the nodes and whitelist `abcd::x` to ensure a line :( .
No FIB entries in both cases.

cgundogan commented:
@BytesGalore I tried it again with desvirt and it worked for me. Are you sure you have the correct version of my branch? I did some fixups the last days that fixed things.

BytesGalore commented:
@cgundogan previously I just failed to configure the white-lists properly :blush:
ACK it does what I think it should do, but I'm not the expert for p2p RPL.

OlegHahm commented:
I think @emmanuelsearch was okay, @BytesGalore did ACK and this is a new feature: Let's go!

cgundogan commented:
finally! thanks for reviewing @BytesGalore and @emmanuelsearch (:

OlegHahm commented:
Hooray! :8ball:

emmanuelsearch commented:
Congrats!

BytesGalore commented:
Yay :D

BytesGalore commented:
hey this might be interesting:
https://www.ietf.org/id/draft-satish-6tisch-aodv-rpl-00.txt

( smells like competition ;) )
